src/Stan/Analysis/Analyser.hs

```diff
@@ -83,6 +84,21 @@ analyseBigTuples insId = matchAstWith isBigTuple insId tuple
         _:_:_:_:_  -> True
         _lessThan4 -> False
 
+{- | Check for usage of the 'String' type in the following places:
+
+* Type signatures: foo :: a -> String
+* Data type fields: newtype HasString = HasString { foo :: String }
+-}
+analyseStringUsage
+    :: Id Inspection
+    -> HieFile
+    -> HieAST TypeIndex
+    -> State VisitorState ()
+analyseStringUsage insId = matchAstWith usesString insId type_
+  where
+    usesString :: HieAST TypeIndex -> Bool
+    usesString = isJust . hieFindIdentifier stringNameMeta
```

Comment:
I see that the test, when `String` is the first argument, is triggering the inspection.
I think this is because of this line. We need to check that it is either in uses String in the data type, or that the function returns it. It is going to be a bit more complicated check though in here. We can start with one check, for example with String in data types, and then deal with the returning String case 🙂

Reply:
Doesn't it make sense to warn agains all uses of `String`? I want to understand the logic behind only checkin the _last usage of String_ in the type signature. 😉

Reply:
Not always you can control the input of all functions, and lots of the functionality in the wild use `Strings`. But when we have custom functions, that we can control, it is recommended to move toward `Text` and return `Text`, that is the meaning for the last argument 🙂